File: myExpenses/src/main/java/org/totschnig/myexpenses/dialog/SortUtilityDialogFragment.java
package org.totschnig.myexpenses.dialog;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;

import com.annimon.stream.Stream;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.adapter.RecyclerListAdapter;
import org.totschnig.myexpenses.adapter.helper.OnStartDragListener;
import org.totschnig.myexpenses.adapter.helper.SimpleItemTouchHelperCallback;

import java.util.AbstractMap;
import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class SortUtilityDialogFragment extends BaseDialogFragment implements OnStartDragListener, DialogInterface.OnClickListener {
  private static final String KEY_ITEMS = "items";
  private ItemTouchHelper mItemTouchHelper;
  private OnConfirmListener callback;
  private RecyclerListAdapter adapter;

  public interface OnConfirmListener {
    void onSortOrderConfirmed(long[] sortedIds);
  }

  public static SortUtilityDialogFragment newInstance(ArrayList<AbstractMap.SimpleEntry<Long, String>> items) {
    final SortUtilityDialogFragment fragment = new SortUtilityDialogFragment();
    Bundle args = new Bundle(1);
    args.putSerializable(KEY_ITEMS, items);
    fragment.setArguments(args);
    return fragment;
  }

  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    try {
      callback = (OnConfirmListener) context;
    } catch (ClassCastException e) {
      throw new ClassCastException(context.toString()
          + " must implement OnConfirmListener");
    }
  }

  @NonNull
  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    AlertDialog.Builder builder = initBuilder();
    Bundle args = savedInstanceState != null ? savedInstanceState : getArguments();
    adapter = new RecyclerListAdapter( this, (ArrayList<AbstractMap.SimpleEntry<Long, String>>) args.getSerializable(KEY_ITEMS));

    RecyclerView recyclerView = new RecyclerView(builder.getContext());
    recyclerView.setHasFixedSize(true);
    recyclerView.setAdapter(adapter);
    recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

    ItemTouchHelper.Callback callback = new SimpleItemTouchHelperCallback(adapter);
    mItemTouchHelper = new ItemTouchHelper(callback);
    mItemTouchHelper.attachToRecyclerView(recyclerView);
    return builder.setTitle(R.string.sort_order)
        .setPositiveButton(android.R.string.ok, this)
        .setNegativeButton(android.R.string.cancel, null)
        .setView(recyclerView)
        .create();
  }

  @Override
  public void onStartDrag(RecyclerView.ViewHolder viewHolder) {
    mItemTouchHelper.startDrag(viewHolder);
  }

  @Override
  public void onClick(DialogInterface dialog, int which) {
    callback.onSortOrderConfirmed(Stream.of(adapter.getItems()).mapToLong(AbstractMap.SimpleEntry::getKey).toArray());
  }

  @Override
  public void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putSerializable(KEY_ITEMS, adapter.getItems());
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/dialog/SortUtilityDialogFragment.kt
package org.totschnig.myexpenses.dialog

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import androidx.core.os.BundleCompat
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.adapter.RecyclerListAdapter
import org.totschnig.myexpenses.adapter.SortableItem
import org.totschnig.myexpenses.adapter.helper.OnStartDragListener
import org.totschnig.myexpenses.adapter.helper.SimpleItemTouchHelperCallback
import java.util.*

class SortUtilityDialogFragment : BaseDialogFragment(), OnStartDragListener,
    DialogInterface.OnClickListener {
    private lateinit var mItemTouchHelper: ItemTouchHelper
    private lateinit var callback: OnConfirmListener
    private lateinit var adapter: RecyclerListAdapter

    interface OnConfirmListener {
        fun onSortOrderConfirmed(sortedIds: LongArray)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        callback = try {
            context as OnConfirmListener
        } catch (_: ClassCastException) {
            throw ClassCastException(
                context.toString()
                        + " must implement OnConfirmListener"
            )
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = initBuilder()
        val args = savedInstanceState ?: requireArguments()
        adapter = RecyclerListAdapter(
            this,
            BundleCompat.getParcelableArrayList(args, KEY_ITEMS, SortableItem::class.java)!!
        )
        val recyclerView = RecyclerView(builder.context)
        recyclerView.setHasFixedSize(true)
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(activity)
        val callback: ItemTouchHelper.Callback = SimpleItemTouchHelperCallback(adapter)
        mItemTouchHelper = ItemTouchHelper(callback)
        mItemTouchHelper.attachToRecyclerView(recyclerView)
        return builder.setTitle(R.string.sort_order)
            .setPositiveButton(android.R.string.ok, this)
            .setNegativeButton(android.R.string.cancel, null)
            .setView(recyclerView)
            .create()
    }

    override fun onStartDrag(viewHolder: RecyclerView.ViewHolder) {
        mItemTouchHelper.startDrag(viewHolder)
    }

    override fun onClick(dialog: DialogInterface, which: Int) {
        callback.onSortOrderConfirmed(adapter.items.map { it.id }.toLongArray())
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(KEY_ITEMS, adapter.items)
    }

    companion object {
        private const val KEY_ITEMS = "items"
        @JvmStatic
        fun newInstance(items: ArrayList<SortableItem>): SortUtilityDialogFragment {
            val fragment = SortUtilityDialogFragment()
            val args = Bundle(1)
            args.putParcelableArrayList(KEY_ITEMS, items)
            fragment.arguments = args
            return fragment
        }
    }
}